File: b4a/b4a.py
import unittest, os, array, sys, subprocess

# opcode list and helper functions

ops = {
  "nop"   : 0,   "lit"   : 1,   "dup"   : 2,    "drop"  : 3,
  "swap"  : 4,   "push"  : 5,   "pop"   : 6,    "loop"  : 7,
  "jump"  : 8,   ";"     : 9,   "<jump" : 10,   ">jump" : 11,
  "!jump" : 12,  "=jump" : 13,  "@"     : 14,   "!"     : 15 ,
  "+"     : 16,  "-"     : 17,  "*"     : 18,   "/mod"  : 19,
  "and"   : 20,  "or"    : 21,  "xor"   : 22,   "<<"    : 23,
  ">>"    : 24,  "0;"    : 25,  "1+"    : 26,   "1-"    : 27,
  "in"    : 28,  "out"   : 29,  "wait"  : 30,
}

def is_int( x ):
  res = True
  try: int( x )
  except ValueError: res = False
  return res

def trim( s ):
  """
  strips leading indentation from a multi-line string.
  for saving bandwidth while making code look nice
  """
  lines = string.split( s, "\n" )
  # strip leading blank line
  if lines[0] == "": lines = lines[ 1: ]
  # strip indentation
  indent = len( lines[ 0 ]) - len( lines[ 0 ].lstrip( ))
  for i in range( len( lines )): lines[ i ] = lines[ i ][ indent: ]
  return '\n'.join( lines )


# assembler

def assemble( src ):
  """
  :: str -> [ int ] or raise ValueError
  --- example program ------------------

    # count until stack overflows
      0           # emits 1 0   ( 1 is an implicit 'lit' opcode )
    :COUNTER
      dup
      inc
      jump COUNTER

  --- end example ----------------------
  """
  labels = {}
  res  = []
  # we do two passes so that we can use labels before they're defined
  for phase in [ 1 , 2 ]:
    here = 0
    for ln, line in enumerate( src.split( "\n" )):
      for word in line.split( ):
        if word.startswith( "#" ): break
        if word.startswith( ":" ): labels[ word[ 1: ]] = here
        else:
          here += 1
          if is_int( word ) :
            here += 1 # add another because of the implicit 'lit' op
            if phase == 2 : res.extend([ ops[ 'lit' ], int( word ) ])
          # the rest are all emitters and we only do that in phase 2:
          elif phase == 1 : pass
          elif word in labels : res.append( labels[ word ])
          elif word in ops : res.append( ops[ word ])
          else : raise ValueError( "unresolved symbol '%s' on line %i"
                                   % ( word, ln + 1 ))
  return res


# save_image ( and 32-bit int encoding )

INT32_FMT = ''
for ch in 'hil':
  if array.array( ch ).itemsize == 4: INT32_FMT = ch
if not INT32_FMT:
  print "couldn't find a 32-bit int in your python's array module."
  print "sorry -- you'll need to update the save_image function or"
  print "use a differnt python version."
  print
  print "see http://docs.python.org/library/array.html#module-array"
  sys.exit()


def save_image( ints, path ):
  """
  write a list of ints as 32-bit binary values
  """
  assert len( ints ), "no instructions given"
  arr = array.array( INT32_FMT, ints )
  img = open( path, 'wb' )
  arr.tofile( img )
  img.close( )



# show control codes

ESC = chr( 27 )
USE_COLOR = True

def maybecolor( string, fg=5, bg=0 ):
  # http://en.wikipedia.org/wiki/ANSI_escape_code#Colors
  if USE_COLOR: return "\x1B[0;3%i;4%im%s\x1B[0m" % ( fg, bg, string )
  else: return string


def catv( string, prefix='^' ):
  """show control characters, like unix cat -v"""
  result = []
  for ch in string:
    if ord( ch ) < 32 and ch not in "\t\n":
      result.append( maybecolor( prefix + '@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_'[ ord( ch )]))
    else: result.append( ch )
  return ''.join( result )

# run

FSEP, GSEP = chr( 28 ), chr( 29 ) # ascii file and group separators

def run( src, imgpath = NGARO_IMG ):
  """
  run the ngaro interpreter and investigate its state.
  see module docstring at top of file for dump format
  """

  save_image( assemble( src ), imgpath )

# test suite

class AssemblerTest( unittest.TestCase ):
  """self-tests for the assembler"""
  def test_labels( self ):
    self.assertEquals( [ 0 ], assemble( ':zero zero' ))
    self.assertEquals( [ 1 ], assemble( 'one :one' ))
  def test_literal( self ):
    """make sure literals increment label counter by 2"""
    self.assertEquals( [ 0, 1, 2, 3 ], assemble( ':zero zero 2 :three three' ))

if __name__=="__main__":
  for arg in sys.argv:
    if os.path.exists( arg ):
      print( arg )

File: b4a/b4a.js
/**
 * Assembler for Retro's ngaro virtual machine (javascript edition)
 *
 * http://retroforth.org/docs/The_Ngaro_Virtual_Machine.html
 *
 * Copyright (c) 2013 Michal J Wallace
 *
 * Permission to use, copy, modify, and/or distribute this software for
 * any purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all copies.
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *
 * Terminology:
 * ----------------------
 * "TOS" is the [T]op [O]f [S]tack
 * "NOS" is [N]ext [O]n [S]tack
 *
 * Stack Effect comments:
 * ----------------------
 *  syntax: ( <data before> ⇢ <data after> | <addr before> ⇢ <addr after> )
 *
 * where <data> is the main stack and <addr> is the return address stack.
 * If the address stack is unaffected, it is not show.
 *
 * Both stacks contains only numbers, but numbers can be treated as::
 *
 * a = address (inside the vm)     n = number
 * f = flag (boolean)              p = port
 *
 * Example : ( n a ⇢ f )  means before: TOS=a NOS=n  after: TOS = f
 *
 * Only the items actually used are shown. There may be many other
 * items on the stack below the indicated values.
 *
 */


/**
 * The Ngaro Instruction set.
 *
 * The names are shortened from the official names in the Ngaro
 * documentation, and the symbols used in retro have all been
 * replaced with words.
 *
 * The goal is that B4 assembly language should look more like
 * assembly language than like forth.
 *
 * Ops that end in ":" are jump opcodes, and must be followed
 * by a label name.
 *
 * As a convenience, all decimal numbers are treated as literals,
 * so there is never any need to use the LIT: mnemonic. (Instead
 * of "LIT: 5 PUSH" you can just say "5 PUSH".
 *
 */
var ops = {

  'NOP'    :  0, // (  ⇢   ) "no op" : do nothing
  '<LIT:>' :  1, // (  ⇢ n ) "literal" : put number on the data stack
  'DUP'    :  2, // ( n ⇢ n ) "duplicate" : copy the TOS
  'DROP'   :  3, // ( n ⇢ ) "drop" : remove the TOS
  'SWAP'   :  4, // (  ⇢ b a ) "swap" : transpose TOS and NOS
  'PUSH'   :  5, // ( n ⇢ | ⇢ n ) "push" : move TOS to the return stack
  'POP'    :  6, // ( ⇢ n | n ⇢ ) "pop" : place top item on return stack in TOS
  'NXT:'   :  7, // [TOS > 1]:( n ⇢ n ) [else]:( n ⇢ ) "next" : if tos >= 1, then DEC and JUMP, else DROP ("for" loop)
  'JMP:'   :  8, // ( ⇢ ) "jump" : jump to the address in the next cell
  'RET'    :  9, // ( ⇢ | a ⇢ ) "return" : jump to previously stored address
  'JLT:'   : 10, // ( x y ⇢ ) "jump if less than" : jump to address in next cell if x < y
  'JGT:'   : 11, // ( x y ⇢ ) "jump if greater than" : jump to address in next cell if x > y
  'JNE:'   : 12, // ( x y ⇢ ) "jump if not equal" : jump to address in next cell if x ≠ y
  'JEQ:'   : 13, // ( x y ⇢ ) "jump if equal" : jump to address in next cell if x = y
  'GET'    : 14, // ( a ⇢ n ) "get" : fetch the value stored at address a
  'PUT'    : 15, // ( n a ⇢ ) "put" : store value n at address a
  'ADD'    : 16, // ( n n ⇢ n ) "add" : replace TOS and NOS with their sum (x + y)
  'SUB'    : 17, // ( x y ⇢ n ) "add" : replace TOS and NOS with their difference (x - y)
  'MUL'    : 18, // ( n n ⇢ n ) "mul" : replace TOS and NOS with their product (x * y)
  'DVM'    : 19, // ( x y ⇢ r q ) "dvm" : divide TOS by NOS, leaving remainder and quotient (ex: 7 5 ⇢ 2 1 )
  'AND'    : 20, // ( n n ⇢ n ) "bitwise and"
  'OR'     : 21, // ( n n ⇢ n ) "bitwise inclusive or"
  'XOR'    : 22, // ( n n ⇢ n ) "bitwise exclusive or"
  'SHL'    : 23, // ( n b ⇢ n ) "shift left"  : shift n left by b bits (ex: 1 4 ⇢ 16 )
  'SHR'    : 24, // ( n b ⇢ n ) "shift right" : shift n right by b bits (ex: 65536 12 ⇢ 16 )
  'ZRET'   : 25, // [TOS ≠ 0]:( n ⇢ n ) [TOS = 0]:( n -- | a -- ) "return if zero"
  'INC'    : 26, // ( n ⇢ n ) "increment" : add 1 to TOS
  'DEC'    : 27, // ( n ⇢ n ) "decrement" : subtract 1 from TOS
  'IN'     : 28, // ( p ⇢ n ) "in"  : read the value currently in port p
  'OUT'    : 29, // ( n p ⇢ ) "out" : write value n to port p
  'WAIT'   : 30  // ( varies ) trigger IO after writing to a port with OUT
};

/**
 * Assemble instructions into machine code.
 *
 * @param {string} src - the assembly language source code to compile.
 * @return {Array} - Array of integers representing the compiled code.
 *
 * See test suite below for some example code.
 *
 */
function assemble( src ) {
  var phase, here, lines, words, word, ln, wd, res = [], labels = {};
  for (phase = 1; phase <= 2; ++phase) {
    here = 0;
    lines = src.split("\n");
    for (ln = 0; ln < lines.length; ++ln) {
      words = lines[ln].split(/\s/);
      for (wd=0; wd<words.length; ++wd) {
        word = words[wd];
        if (word[0] == "#") break;
        if (word[0] == ":") labels[ word.substr( 1 )] = here;
        else {
          here++;
          if (/^-?\d+$/.exec( word )) {  // match int literal
            here++; // because of the implicit "LIT" opcode
            if (phase==2) res.push(ops['<LIT:>'], parseInt(word, 10));
          }
          else if (phase == 1) {} // phase 1 only for labels
          else if (word in labels) res.push( labels[ word ]);
          else if (word in ops) res.push( ops[ word ]);
          else throw "Unresolved symbol '" + word + "' on line " + ln + ".";
        }
      }
    }
  }
  return res;
}

//---- test suite ----------------------------------------------

/**
 * Compare two arrays item by item to determine whether the contents are the same.
 * @param a - an array
 * @param b - an array
 * @returns {boolean} true if the items in the two arrays are equal
 */
function same(a, b) {
  for (var i=0; i<a.length; i++) if (a[i] != b[i]) return false;
  return true;
}

function assertSame(a,b, msg) {
  if (!same(a,b)) console.log("failed test <" + msg + "> : " + a + " != " + b);
}

function test_assembler() {

  assertSame(assemble( ':zero zero' ), [ 0 ],
    "normal label use");

  assertSame(assemble( 'one :one'   ), [ 1 ],
    "forward label use"); // the ++ comes first

  assertSame(assemble( ':zero zero 2 :three three' ), [ 0, 1, 2, 3 ],
    "literals should increment label counter by 2" );

}

if (console.clear !== undefined) console.clear(); // only in firebug
console.log("running tests.");
test_assembler();
console.log("done.");
